lib/galaxy/tool_util/cwl/parser.py

```diff
@@ -173,29 +168,29 @@ def label(self):
     def to_persistent_representation(self):
         """Return a JSON representation of this tool. Not for serialization
         over the wire, but serialization in a database."""
-        # TODO: Replace this with some more readable serialization,
-        # I really don't like using pickle here.
-        if PERSISTED_REPRESENTATION == "cwl_tool_object":
-            persisted_obj = remove_pickle_problems(self._tool)
-        else:
-            persisted_obj = self._raw_process_reference
+        persisted_obj = self._tool.tool
+        persisted_obj["requirements"] = self.requirements
+        if not persisted_obj.get("cwlVersion"):
+            # This happens for any inline process, but getting it from metadata is correct for inline processes at least
+            persisted_obj["cwlVersion"] = self._tool.metadata["cwlVersion"]
         return {
             "class": self._class,
-            "pickle": unicodify(base64.b64encode(pickle.dumps(persisted_obj, pickle.HIGHEST_PROTOCOL))),
+            # Should maybe be yaml instead
+            "raw_process_reference": persisted_obj,
             "uuid": self.uuid,
         }
 
     @staticmethod
-    def from_persistent_representation(as_object) -> "ToolProxy":
+    def from_persistent_representation(as_object, strict_cwl_validation=True, tool_directory=None) -> "ToolProxy":
         """Recover an object serialized with to_persistent_representation."""
         if "class" not in as_object:
             raise Exception("Failed to deserialize tool proxy from JSON object - no class found.")
-        if "pickle" not in as_object:
-            raise Exception("Failed to deserialize tool proxy from JSON object - no pickle representation found.")
-        if "uuid" not in as_object:
-            raise Exception("Failed to deserialize tool proxy from JSON object - no uuid found.")
-        to_unpickle = base64.b64decode(as_object["pickle"])
-        loaded_object = pickle.loads(to_unpickle)
+        loaded_object = tool_proxy(
+            tool_object=as_object["raw_process_reference"],
+            strict_cwl_validation=strict_cwl_validation,
+            tool_directory=tool_directory,
+            uuid=as_object.get("uuid"),
```

Comment:
```suggestion
            uuid=as_object["uuid"],
```

Reply:
I am not sure that's what we wanted to do in the downstream branch. `_to_cwl_tool_object` can set the uuid, and I've spent some time in this, I'd rather not change that here.

Reply:
There's definitely work to be done there, I'm not sure uuid is the right thing either, this kind of replaces the tool id as a unique handle, if anything it should be the document hash, but that's also tricky due to commented yaml etc